src/element-service.js

```diff
@@ -85,3 +91,61 @@ function isElementScheduled(win, elementName) {
   }
   return !!win.ampExtendedElements[elementName];
 }
+
+
+/**
+ * Returns a promise for a service for the given id and window. Also expects
+ * an element that has the actual implementation. The promise resolves when
+ * the implementation loaded.
+ * Users should typically wrap this as a special purpose function (e.g.
+ * viewportForDoc(...)) for type safety and because the factory should not be
+ * passed around.
+ * @param {!Node|!./service/ampdoc-impl.AmpDoc} nodeOrDoc
+ * @param {string} id of the service.
+ * @param {string} providedByElement Name of the custom element that provides
+ *     the implementation of this service.
+ * @return {!Promise<*>}
+ */
+export function getElementServiceForDoc(nodeOrDoc, id, providedByElement) {
+  return getElementServiceIfAvailableForDoc(nodeOrDoc, id, providedByElement)
+      .then(service => {
+        return user().assert(service,
+              'Service %s was requested to be provided through %s, ' +
+              'but %s is not loaded in the current page. To fix this ' +
+              'problem load the JavaScript file for %s in this page.',
+              id, providedByElement, providedByElement, providedByElement);
+      });
+}
+
+/**
+ * Same as getElementService but produces null if the given element is not
+ * actually available on the current page.
+ * @param {!Node|!./service/ampdoc-impl.AmpDoc} nodeOrDoc
+ * @param {string} id of the service.
+ * @param {string} providedByElement Name of the custom element that provides
+ *     the implementation of this service.
+ * @return {!Promise<?Object>}
+ */
+export function getElementServiceIfAvailableForDoc(
+    nodeOrDoc, id, providedByElement) {
+  const ampdoc = getAmpdoc(nodeOrDoc);
+  const s = getServicePromiseOrNullForDoc(nodeOrDoc, id);
+  if (s) {
+    return /** @type {!Promise<?Object>} */ (s);
+  }
+  // Microtask is necessary to ensure that window.ampExtendedElements has been
+  // initialized.
+  return Promise.resolve().then(() => {
+    if (isElementScheduled(ampdoc.win, providedByElement)) {
+      return getServicePromiseForDoc(nodeOrDoc, id);
+    }
+    // Wait for HEAD to fully form before denying access to the service.
+    // TODO(mk): DOES THIS NEED TO BE CHANGED?
+    return dom.waitForBodyPromise(ampdoc.getRootNode()).then(() => {
```

Comment:
So, as described elsewhere as well: `waitForBody` is not the right call here. Instead we should just use `ampdoc.whenBodyAvailable`. And yes, you need to wait for body here.